core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -356,4 +374,229 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
+    : List[PortRangeResourceInfo] = {
+    // A resource can have multiple values in the offer since it can either be from
+    // a specific role or wildcard.
+    // Extract role info and port range for every port resource in the offer
+    res.asScala.filter(_.getName == name)
+      .map{res => PortRangeResourceInfo(res.getRole, res.getRanges.getRangeList.asScala
+        .map(r => (r.getBegin, r.getEnd)).toList) }.toList
+  }
+
+  /** Helper method to get a pair of assigned and remaining ports along with role info */
+  private def reservePorts(
+      availablePortRanges: List[PortRangeResourceInfo],
+      wantedPorts: List[Long])
+    : (List[PortRangeResourceInfo], List[PortRangeResourceInfo]) = {
+    if (wantedPorts.isEmpty) { // port list is empty we didnt consume any resources
+      return (availablePortRanges, List())
+    }
+    var tmpLeft = availablePortRanges
+    val tmpRanges = for {port <- wantedPorts}
+      yield {
+        val ret = findPortAndSplitRange(port, tmpLeft)
+        val rangeToRemove = ret._1
+        val diffRanges = tmpLeft.filterNot{r => r == rangeToRemove}
+        val newRangesLeft = diffRanges ++ List(ret._2).flatMap(p => p)
+        tmpLeft = newRangesLeft
+        ret
+      }
+    val rangesToRemove = tmpRanges.map(x => x._1)
+    val newRangesLeft = (availablePortRanges ++ tmpRanges.flatMap{x => x._2})
+      .flatMap{r => removeRanges(r, rangesToRemove)}
+    val newRanges = tmpRanges.map{r => PortRangeResourceInfo(r._1.role, List((r._3, r._3)))}
+    // return a list of the resources left and one with the newly assigned ones
+    // we need to carry the role info as well
+    (newRangesLeft, newRanges)
+  }
+
+  private def removeRanges(
+      rangeA: PortRangeResourceInfo,
+      rangesToRemove: List[PortRangeResourceInfo]): Option[PortRangeResourceInfo] = {
+    val ranges = rangeA.value.filterNot(rangesToRemove.flatMap{_.value}.toSet)
+    if (ranges.isEmpty) {
+      None
+    } else {
+      Some(PortRangeResourceInfo(rangeA.role, ranges))
+    }
+  }
+
+  private def createMesosPortResource(
+      ranges: List[(Long, Long)],
+      role: Option[String] = None): List[Resource] = {
+    ranges.map { range =>
+      val rangeValue = Value.Range.newBuilder()
+      rangeValue.setBegin(range._1)
+      rangeValue.setEnd(range._2)
+      val builder = Resource.newBuilder()
+        .setName("ports")
+        .setType(Value.Type.RANGES)
+        .setRanges(Value.Ranges.newBuilder().addRange(rangeValue))
+      role.foreach { r => builder.setRole(r) }
+      builder.build()
+    }
+  }
+
+  private def pickRandomPortsFromRanges(
+      ranges: List[PortRangeResourceInfo],
+      numToPick: Int): List[Long] = {
+    if (numToPick == 0) {
+      return List()
+    }
+    val ports = scala.util.Random.
+      shuffle(ranges.flatMap(p => p.value.flatMap(r => (r._1 to r._2).toList))
+      .distinct
+    )
+    require(ports.size >= numToPick)
```

Comment:
ok i will fix it but i am sure i have seen similar restrictions elsewhere, thats why i followed the pattern.